File: frontend/src/api/knowledge.ts
import { API_BASE_URL } from "./ingestion";

export type EntityType =
  | "video"
  | "creator"
  | "transcript"
  | "screenshot"
  | "claim"
  | "source"
  | "paper"
  | "author"
  | "evidence"
  | "report"
  | "rating"
  | "topic";

export type ConsistencyStatus = "synced" | "drift" | "broken";
export type RatingBadgeText =
  | "strong evidence history"
  | "mixed evidence history"
  | "limited evidence"
  | "insufficient history";

export type KnowledgeSection = {
  name: string;
  entity_type?: EntityType | null;
  count: number;
};

export type KnowledgeNoteListItem = {
  uuid: string;
  entity_type: EntityType;
  title: string;
  vault_path: string;
  slug?: string | null;
  excerpt: string;
  updated_at?: string | null;
};

export type KnowledgeRelationshipView = {
  uuid?: string | null;
  relationship_type: string;
  source_uuid?: string | null;
  target_uuid?: string | null;
  source_title?: string | null;
  target_title?: string | null;
  direction?: string | null;
  provenance?: string | null;
  vault_path?: string | null;
};

export type KnowledgeBacklink = {
  uuid: string;
  entity_type: EntityType;
  title: string;
  vault_path: string;
  relationship_type?: string | null;
};

export type KnowledgeAnnotation = {
  uuid: string;
  target_entity_uuid: string;
  author: string;
  body: string;
  created_at: string;
  updated_at: string;
};

export type KnowledgeRatingSummary = {
  rating_uuid?: string | null;
  target_uuid: string;
  target_entity_type: EntityType;
  title?: string | null;
  vault_path?: string | null;
  badge: RatingBadgeText;
  experimental: boolean;
  evidence_count: number;
  label_distribution: Record<string, number>;
  source_basis: string[];
  confidence_level: string;
  report_version_uuids: string[];
};

export type KnowledgeConsistencyIssue = {
  status: ConsistencyStatus;
  affected_uuid?: string | null;
  entity_type?: EntityType | null;
  title?: string | null;
  vault_path?: string | null;
  store: string;
  issue: string;
  suggested_repair: string;
};

export type KnowledgeConsistencySummary = {
  status: ConsistencyStatus;
  checked_notes: number;
  missing_mongo_records: number;
  missing_qdrant_payloads: number;
  broken_relationships: number;
  orphan_vectors: number;
  issues: KnowledgeConsistencyIssue[];
};

export type KnowledgeNoteDetail = {
  uuid: string;
  entity_type: EntityType;
  title: string;
  vault_path: string;
  slug?: string | null;
  frontmatter: Record<string, unknown>;
  body_markdown: string;
  wiki_links: string[];
  relationships: KnowledgeRelationshipView[];
  backlinks: KnowledgeBacklink[];
  annotations: KnowledgeAnnotation[];
  rating?: KnowledgeRatingSummary | null;
  consistency?: KnowledgeConsistencySummary | null;
  updated_at?: string | null;
};

export type KnowledgeSearchResult = {
  uuid: string;
  entity_type: EntityType;
  title: string;
  vault_path: string;
  snippet: string;
  score?: number | null;
  source?: string | null;
  chunk_id?: string | null;
  source_uuid?: string | null;
  relationship_uuid?: string | null;
  relationship_uuids: string[];
  vector_backed: boolean;
};

export type KnowledgeGraphNode = {
  uuid: string;
  entity_type: EntityType;
  title: string;
  vault_path?: string | null;
  degree: number;
};

export type KnowledgeGraphEdge = {
  uuid: string;
  relationship_type: string;
  source_uuid: string;
  target_uuid: string;
  provenance: string;
  direction: string;
};

export type KnowledgeGraphResponse = {
  selected_uuid: string;
  nodes: KnowledgeGraphNode[];
  edges: KnowledgeGraphEdge[];
  important_nodes: KnowledgeGraphNode[];
  clusters: Record<string, string[]>;
};

export type RatingRecord = {
  rating_uuid: string;
  target_uuid: string;
  target_entity_type: "creator" | "paper" | "author" | "source";
  target_title: string;
  badge: RatingBadgeText;
  experimental: boolean;
  evidence_count: number;
  label_distribution: Record<string, number>;
  source_basis: string[];
  confidence_level: "low" | "medium" | "high";
  report_version_uuids: string[];
  evidence_uuids: string[];
  relationship_uuids: string[];
  generated_at: string;
  vault_path: string;
};

async function parseResponse<T>(response: Response, fallbackError: string): Promise<T> {
  if (!response.ok) {
    const payload = await response.json().catch(() => ({}));
    throw new Error(typeof payload.detail === "string" ? payload.detail : fallbackError);
  }
  return response.json() as Promise<T>;
}

export async function fetchKnowledgeSections(): Promise<KnowledgeSection[]> {
  return parseResponse<KnowledgeSection[]>(
    await fetch(`${API_BASE_URL}/knowledge/sections`),
    "knowledge_sections_fetch_failed",
  );
}

export async function fetchKnowledgeNotes(entityType?: EntityType): Promise<KnowledgeNoteListItem[]> {
  const suffix = entityType ? `?entity_type=${encodeURIComponent(entityType)}` : "";
  return parseResponse<KnowledgeNoteListItem[]>(
    await fetch(`${API_BASE_URL}/knowledge/notes${suffix}`),
    "knowledge_notes_fetch_failed",
  );
}

export async function fetchKnowledgeNote(noteUuid: string): Promise<KnowledgeNoteDetail> {
  return parseResponse<KnowledgeNoteDetail>(
    await fetch(`${API_BASE_URL}/knowledge/notes/${noteUuid}`),
    "knowledge_note_fetch_failed",
  );
}

export async function createKnowledgeAnnotation(
  noteUuid: string,
  body: string,
): Promise<KnowledgeAnnotation> {
  return parseResponse<KnowledgeAnnotation>(
    await fetch(`${API_BASE_URL}/knowledge/notes/${noteUuid}/annotations`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ body }),
    }),
    "knowledge_annotation_create_failed",
  );
}

export async function searchKnowledge(
  query: string,
  entityTypes: EntityType[] = [],
): Promise<{ query: string; results: KnowledgeSearchResult[]; count: number; vector_backed: boolean }> {
  const params = new URLSearchParams({ q: query });
  entityTypes.forEach((entityType) => params.append("entity_type", entityType));
  return parseResponse(
    await fetch(`${API_BASE_URL}/knowledge/search?${params.toString()}`),
    "knowledge_search_failed",
  );
}

export async function fetchKnowledgeGraph(noteUuid: string): Promise<KnowledgeGraphResponse> {
  return parseResponse<KnowledgeGraphResponse>(
    await fetch(`${API_BASE_URL}/knowledge/graph/${noteUuid}`),
    "knowledge_graph_fetch_failed",
  );
}

export async function fetchKnowledgeConsistency(): Promise<KnowledgeConsistencySummary> {
  return parseResponse<KnowledgeConsistencySummary>(
    await fetch(`${API_BASE_URL}/knowledge/consistency`),
    "knowledge_consistency_fetch_failed",
  );
}

export async function fetchRating(targetUuid: string): Promise<RatingRecord> {
  return parseResponse<RatingRecord>(
    await fetch(`${API_BASE_URL}/knowledge/ratings/${targetUuid}`),
    "knowledge_rating_fetch_failed",
  );
}

export async function refreshRating(targetUuid: string): Promise<RatingRecord> {
  return parseResponse<RatingRecord>(
    await fetch(`${API_BASE_URL}/knowledge/ratings/${targetUuid}/refresh`, {
      method: "POST",
    }),
    "knowledge_rating_refresh_failed",
  );
}

File: backend/app/schemas/knowledge.py
from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from app.schemas.entities import EntityType


class ConsistencyStatus(str, Enum):
    synced = "synced"
    drift = "drift"
    broken = "broken"


class RatingBadge(str, Enum):
    strong_evidence_history = "strong evidence history"
    mixed_evidence_history = "mixed evidence history"
    limited_evidence = "limited evidence"
    insufficient_history = "insufficient history"


class KnowledgeSection(BaseModel):
    name: str
    entity_type: EntityType | None = None
    count: int


class KnowledgeNoteListItem(BaseModel):
    uuid: UUID
    entity_type: EntityType
    title: str
    vault_path: str
    slug: str | None = None
    excerpt: str = ""
    updated_at: datetime | None = None


class KnowledgeRelationshipView(BaseModel):
    uuid: UUID | None = None
    relationship_type: str
    source_uuid: UUID | None = None
    target_uuid: UUID | None = None
    source_title: str | None = None
    target_title: str | None = None
    direction: str | None = None
    provenance: str | None = None
    vault_path: str | None = None


class KnowledgeBacklink(BaseModel):
    uuid: UUID
    entity_type: EntityType
    title: str
    vault_path: str
    relationship_type: str | None = None


class KnowledgeAnnotation(BaseModel):
    uuid: UUID = Field(default_factory=uuid4)
    target_entity_uuid: UUID
    author: str = "user"
    body: str
    created_at: datetime
    updated_at: datetime


class KnowledgeAnnotationCreate(BaseModel):
    body: str


class KnowledgeRatingSummary(BaseModel):
    rating_uuid: UUID | None = None
    target_uuid: UUID
    target_entity_type: EntityType
    title: str | None = None
    vault_path: str | None = None
    badge: RatingBadge
    experimental: bool
    evidence_count: int
    label_distribution: dict[str, int] = Field(default_factory=dict)
    source_basis: list[str] = Field(default_factory=list)
    confidence_level: str
    report_version_uuids: list[UUID] = Field(default_factory=list)


class KnowledgeConsistencyIssue(BaseModel):
    status: ConsistencyStatus
    affected_uuid: UUID | None = None
    entity_type: EntityType | None = None
    title: str | None = None
    vault_path: str | None = None
    store: str
    issue: str
    suggested_repair: str


class KnowledgeConsistencySummary(BaseModel):
    status: ConsistencyStatus
    checked_notes: int = 0
    missing_mongo_records: int = 0
    missing_qdrant_payloads: int = 0
    broken_relationships: int = 0
    orphan_vectors: int = 0
    issues: list[KnowledgeConsistencyIssue] = Field(default_factory=list)


class KnowledgeNoteDetail(BaseModel):
    uuid: UUID
    entity_type: EntityType
    title: str
    vault_path: str
    slug: str | None = None
    frontmatter: dict[str, Any] = Field(default_factory=dict)
    body_markdown: str
    wiki_links: list[str] = Field(default_factory=list)
    relationships: list[KnowledgeRelationshipView] = Field(default_factory=list)
    backlinks: list[KnowledgeBacklink] = Field(default_factory=list)
    annotations: list[KnowledgeAnnotation] = Field(default_factory=list)
    rating: KnowledgeRatingSummary | None = None
    consistency: KnowledgeConsistencySummary | None = None
    updated_at: datetime | None = None


class KnowledgeSearchResult(BaseModel):
    uuid: UUID
    entity_type: EntityType
    title: str
    vault_path: str
    snippet: str
    score: float | None = None
    source: str | None = None
    chunk_id: str | None = None
    source_uuid: UUID | None = None
    relationship_uuid: UUID | None = None
    relationship_uuids: list[UUID] = Field(default_factory=list)
    vector_backed: bool = False


class KnowledgeGraphNode(BaseModel):
    uuid: UUID
    entity_type: EntityType
    title: str
    vault_path: str | None = None
    degree: int = 0


class KnowledgeGraphEdge(BaseModel):
    uuid: UUID
    relationship_type: str
    source_uuid: UUID
    target_uuid: UUID
    provenance: str
    direction: str


class KnowledgeGraphResponse(BaseModel):
    selected_uuid: UUID
    nodes: list[KnowledgeGraphNode] = Field(default_factory=list)
    edges: list[KnowledgeGraphEdge] = Field(default_factory=list)
    important_nodes: list[KnowledgeGraphNode] = Field(default_factory=list)
    clusters: dict[str, list[UUID]] = Field(default_factory=dict)
